staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/algorithm.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package objectmeta
+
+import (
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// Coerce checks types of embedded ObjectMeta and TypeMeta and prunes unknown fields inside the former.
+// It does coerce ObjectMeta and TypeMeta at the root if root is true.
+func Coerce(pth *field.Path, obj interface{}, s *structuralschema.Structural, root bool) *field.Error {
+	if root {
+		if s == nil {
+			s = &structuralschema.Structural{}
+		}
+		clone := *s
+		clone.XEmbeddedResource = true
+		s = &clone
+	}
+	return coerce(pth, obj, s)
+}
+
+func coerce(pth *field.Path, x interface{}, s *structuralschema.Structural) *field.Error {
+	if s == nil {
+		return nil
+	}
+	switch x := x.(type) {
+	case map[string]interface{}:
+		for k, v := range x {
+			if s.XEmbeddedResource {
+				switch k {
+				case "apiVersion", "kind":
+					if _, ok := v.(string); !ok {
+						return field.Invalid(pth, v, "must be a string")
+					}
+				case "metadata":
+					meta, _, err := GetObjectMeta(x, true)
```

Comment:
@liggitt the true means to drop invalid fields (you remember that we go field by field and do partial json.Unmarshal). I tend to make this strict instead of dropping field. Wdyt?

Reply:
What is the consequence if they add x-kubernetes-embedded-resource to an existing CRD with existing persisted data that fails this? Does that result in a decode error that makes a list of custom resource instances fail with a server error? If so, that sounds problematic

Reply:
Yes, that will be the case.

Dropping fields of wrong type is also kind of strange. But it looks like we will never get rid of that again.

Reply:
Following the same logic as for the root ObjectMeta: strict for request objects, non-strict on etcd read.